rfd/0XYZ-nested-accesslists.md

```diff
@@ -0,0 +1,121 @@
+---
+title: RFD 0XYZ - Nested Access lists
+authors: Alex McGrath (alex.mcgrath@goteleport.com)
+state: in development
+---
+
+# Related
+  * Partial implementation PR: https://github.com/gravitational/teleport/pull/38738
+
+# What
+
+This RFD proposes adding support for including owners and
+members in access lists from other access lists
+
+# Why
+
+This allows a hierarchical access list structure for organizations
+which is supported by access list implementations being integrated
+with, for example Azure Entra supports including groups in other
+groups which we want to be able to mimic the structure of in Teleport
+access lists.
+
+Users in an access list hierarchy will inherit the granted roles and
+traits for members and owners from access lists referencing the lists
+they're in.
+
+# Implementations
+
+New fields will be introduced into the access_list type:
+
+```yaml
+kind: access_list
+metadata:
+  name: e69aa529-2a7f-4de2-87be-97fd94309a9f
+spec:
+  grants:
+    roles:
+    - access
+    - auditor
+    traits: {}
+  # list of references to other access lists, for users to include in this access list
+  member_access_lists:
+  - name: ea4cbbc7-bee1-49b3-bf78-734b4b27ea38
+  # list of references to other access lists, for owners to include in this access list
+  owner_access_lists:
+  - name: 3e9df1e7-0b8a-4984-b2e8-5bc0d7b356a9
+  title: access-list-a
+version: v1
+```
+
+These fields will contain a list of other access lists that will be
+included in the access list.
+
+# Implementation considerations
+
+The implementation will not support looping within the heirarchy as
```

Comment:
I'm not sure if I understand this point. What looping is? I'd assume that an access list would fetch all parents, merge them into one set of rules an run all checks against it.
Also, are we going to support cycles? If no, how are we going to prevent them?

Reply:
I'm reading this "looping" as a synonym for cycles.

Reply:
I don't think that looping has anything to do with cycles. `for(i = 0; i < 10; i++)` and `while(true)` are both loops where the first one will end and the second will loop forever.
All graph algorithms terminate on DAG, but a cycle make cause an algorithm to never terminate.
@justinas AFAIR you mentioned that Entra allows cycles in the setup? If this true, then we won't be able to import all roles.

Reply:
> @justinas AFAIR you mentioned that Entra allows cycles in the setup? If this true, then we won't be able to import all roles.

That's true, Entra does not limit you from making cycles `A -> B -> A` from groups. IMO it should not be a common pattern, so I'd be fine with not supporting importing such cycles, but that would technically mean we are not producing an accurate view of the directory.

Reply:
I meant looping in terms of a cycle in a graph, just misnamed it here. We just want to avoid it I think to avoid introducing overly complicated configuration options, with the hope that they're not too common, if they are required by somone, it shouldnt be too difficult to implement it